[webOS] Fix timezone reading

sundermann opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Timezone was not set correctly on startup on webOS. Some refactoring is done first to avoid duplicating code. The diff is a bit large because `clang format` required fixing up the header inclusion order and variable spacing

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
The webOS jailer only mounts `/etc/localtime` which is a symlink to `/var/luna/preferences/localtime`. `/var/luna/preferences/localtime ` however is another symlink to the actual zone info.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
webOS22/7

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Fixes timezone initialization

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [x] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

sundermann committed:
[PosixTimezone] Refactor timezone reading

sundermann commented:
It seems `std::filesystem` is not available until macOS 10.15 and `fstream` is not available until NDK 24.

throwaway96 commented:
Information for webOS 4.1.0 (MStar M3 SoC, using the YouTube jail as an example):

`/var/luna/preferences/timezone` is present in the jail, but it's `Universal` instead of the system timezone (`America/New_York`):
```
root@lg-m3[m3 4.1.0]:~# cat /var/palm/jail/youtube.leanback.v4/var/luna/preferences/timezone
Universal
```

`/etc/localtime` is a symlink to `/var/luna/preferences/localtime`:
```
root@lg-m3[m3 4.1.0]:~# readlink /var/palm/jail/youtube.leanback.v4/etc/localtime
/var/luna/preferences/localtime
```

`/var/luna/preferences/localtime` (symlink to correct zoneinfo) is present in the jail:
```
root@lg-m3[m3 4.1.0]:~# readlink /var/palm/jail/youtube.leanback.v4/var/luna/preferences/localtime
/usr/share/zoneinfo/America/New_York
```

It seems that `/usr/share/zoneinfo` is available from inside the jail, e.g.:
```
  File: /var/palm/jail/youtube.leanback.v4/usr/share/zoneinfo/America/New_York
  Size: 3545            Blocks: 7          IO Block: 1024   regular file
Device: b31dh/45853d    Inode: 61542       Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2018-09-17 22:29:54.000000000
Modify: 2018-09-17 22:29:54.000000000
Change: 2018-09-17 22:29:54.000000000
```

These are the relevant lines in `/etc/jail_native.conf` (although I'm not sure exactly which configuration file is used, or how it's chosen, these lines seem to be common to all):
```
# /var/luna/preferences
mkdir /var/luna
mkdir /var/luna/preferences
mount ro /var/luna/preferences

link /var/luna/preferences/localtime /etc/localtime
```

`/usr/share` (apparently responsible for the existence of `/usr/share/zoneinfo`, although they mount `/usr/share/fonts` explicitly) is also mounted by `jail_native.conf`:
```
mount ro /usr/share
```

These files are the same outside of the jail. It looks like `/var/luna/preferences/timezone` has the wrong timezone setting. That may be an artifact of the strange way my webOS boards are configured (missed a firstuse step?), but I can look into it on other boards/versions.

sundermann commented:
Thank you for checking this on different webOS versions. As you suggested I have changed to reading the symlink path from `/var/luna/preferences/localtime` instead

wsnipex commented on xbmc/platform/posix/PosixTimezone.cpp:
```diff
@@ -195,44 +197,65 @@ void CPosixTimezone::SetTimezone(const std::string& timezoneName)
-   }
+  // RHEL and maybe other distros make /etc/localtime a symlink
+  if (timezoneName.empty())
+    timezoneName = ReadFromLocaltime("/etc/localtime");
```
this is how systemd configures the timezone, so it should work on any distro with systemd.
I'd make this the first option we try

sundermann replied:
`/etc/localtime` points to `/var/luna/preferences/localtime` on webOS.  So we'd retrieve `preferences/localtime` if it was tried first

wsnipex replied:
don't you already use readlink? And that's what was used before as well

sundermann replied:
Oh sorry, I meant `realpath`

wsnipex replied:
that should work as well. Quickly tested on ubuntu

sundermann replied:
Do you think it's better to change only to `realpath` and keep the rest of the code or also apply the refactoring as well and adapting that to use `realpath` too?

sundermann committed:
[PosixTimezone][webOS] Use realpath to resolve localtime path

On webOS /etc/localtime is a symlink to /var/luna/preferences/localtime which in turn is another symlink to the real timezone file. In this case the timezone file must be resolved twice which realpath does

wsnipex reviewed (approved):
looks good to me
